### PR TITLE
Only save login credentials in database once there is an external storage that needs it

### LINKED ISSUE #6343: PostgreSQL : duplicate key value violates unique constraint error message
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps to reproduce
just logon on nextcloud and check syslog/postgresql log messages

### Server configuration
Nextcloud version : 12.0.2
Operating system and version : archlinux (current)
Apache or nginx version : nginx 1.12.1
PHP version : 7.1.7
PostgreSQL : 9.6.3

Hi all !

I encounter a postgresql error message in my syslog everytime i login on my nextcloud instance :

ERROR: duplicate key value violates unique constraint "oc_credentials_pkey"
DETAIL: Key (“user”, identifier)=(myuser, password::logincredentials/credentials) already exists.
STATEMENT: INSERT INTO “oc_credentials” (“user”, “identifier”, “credentials”) VALUES($1, $2, $3)

No problem for login/logout and using nextcloud, only this error message.

Thanks !

### PR DESCRIPTION
**Pull request by icewind1991**
Also fixes #6343

### CODE REVIEW

**Comment by kesselb** [comment]
This patch also helps with https://github.com/nextcloud/server/issues/6343 as the data is not written anymore on login. 

How do you detect changed credentials without listening to postLogin?

**Comment by icewind1991** [comment]
> How do you detect changed credentials without listening to postLogin?

Forgot about that :see_no_evil: , will need to add something for that

**Comment by icewind1991** [comment]
> > How do you detect changed credentials without listening to postLogin?
> 
> Forgot about that see_no_evil , will need to add something for that

done

**Comment by kesselb** [comment]
> done

Nice :+1: Much cleaner than before and will still help with https://github.com/nextcloud/server/issues/6343 as the password is only updated on change :rocket:

File: apps/files_external/lib/BackgroundJob/CredentialsCleanup.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+		IUserManager $userManager
+	) {
+		parent::__construct($time);
+
+		$this->credentialsManager = $credentialsManager;
+		$this->userGlobalStoragesService = $userGlobalStoragesService;
+		$this->userManager = $userManager;
+
+		// run every hour
+		$this->setInterval(60 * 60);
```
Comment (MorrisJobke) [inline apps/files_external/lib/BackgroundJob/CredentialsCleanup.php:53]: Is it really needed to be run once an hour? That basically only changes once an admin changes an external storage configuration and on upgrade to a version that includes this code. Or did I miss something? Shouldn't be once a day or even once a week fully enough?

Comment (icewind1991) [reply apps/files_external/lib/BackgroundJob/CredentialsCleanup.php]: switched it to daily, that seems like a decent compromise

**Comment by kesselb** [comment]
@rullzer @MorrisJobke @juliushaertl 🏓

File: apps/files_external/lib/Lib/Auth/Password/LoginCredentials.php [hunk 2]
```diff
@@ -44,45 +49,72 @@ class LoginCredentials extends AuthMechanism {
...
 			->setScheme(self::SCHEME_PASSWORD)
 			->setText($l->t('Log-in credentials, save in database'))
 			->addParameters([
-			])
-		;
+			]);
 
-		\OCP\Util::connectHook('OC_User', 'post_login', $this, 'authenticate');
+		$eventDispatcher->addListener(UserLoggedInEvent::class, [$this, 'updateCredentials']);
+		$eventDispatcher->addListener(PasswordUpdatedEvent::class, [$this, 'updateCredentials']);
```
Comment (MorrisJobke) [inline apps/files_external/lib/Lib/Auth/Password/LoginCredentials.php:68]: 😍

Comment (MorrisJobke) [reply apps/files_external/lib/Lib/Auth/Password/LoginCredentials.php:68]: Would be even better with separate listener classes, but this is already a bit better.

File: apps/files_external/lib/Lib/Auth/Password/LoginCredentials.php [hunk 3]
```diff
@@ -44,45 +49,72 @@ class LoginCredentials extends AuthMechanism {
...
-		];
-		$this->credentialsManager->store($userId, self::CREDENTIALS_IDENTIFIER, $credentials);
+	public function updateCredentials($event) {
+		if ($event instanceof UserLoggedInEvent && $event->isTokenLogin()) {
+			return;
+		}
+
+		$stored = $this->credentialsManager->retrieve($event->getUser()->getUID(), self::CREDENTIALS_IDENTIFIER);
+
+		if ($stored && $stored['password'] != $event->getPassword()) {
```
Comment (MorrisJobke) [inline apps/files_external/lib/Lib/Auth/Password/LoginCredentials.php:81]:
```suggestion
		if ($stored && $stored['password'] !== $event->getPassword()) {
```

**Comment by MorrisJobke** [comment]
Let me fix this :)

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Review by blizzz — APPROVED** [review]

**Pull request merged**